libraries/GCS_MAVLink/GCS_FTP.cpp

```diff
@@ -96,50 +90,34 @@ void GCS_MAVLINK::handle_file_transfer_protocol(const mavlink_message_t &msg) {
     }
 }
 
-void GCS_MAVLINK::send_ftp_replies(void)
+bool GCS_MAVLINK::send_ftp_reply(const pending_ftp &reply)
 {
     /*
       provide same banner we would give with old param download
     */
-    if (ftp.need_banner_send_mask & (1U<<chan)) {
-        ftp.need_banner_send_mask &= ~(1U<<chan);
+    if (ftp.need_banner_send_mask & (1U<<reply.chan)) {
+        ftp.need_banner_send_mask &= ~(1U<<reply.chan);
         send_banner();
     }
-    
-    if (ftp.replies == nullptr || ftp.replies->is_empty()) {
-        return;
-    }
-
-    for (uint8_t i = 0; i < 20; i++) {
-        if (!HAVE_PAYLOAD_SPACE(chan, FILE_TRANSFER_PROTOCOL)) {
-            return;
-        }
-        if ((i > 0) && comm_get_txspace(chan) < (2 * (packet_overhead() + MAVLINK_MSG_ID_FILE_TRANSFER_PROTOCOL_LEN))) {
-            // if this isn't the first packet we have to leave deadspace for the next message
-            return;
-        }
-
-        struct pending_ftp reply;
-        uint8_t payload[251] = {};
-        if (ftp.replies->peek(reply) && (reply.chan == chan)) {
-            put_le16_ptr(payload, reply.seq_number);
-            payload[2] = reply.session;
-            payload[3] = static_cast<uint8_t>(reply.opcode);
-            payload[4] = reply.size;
-            payload[5] = static_cast<uint8_t>(reply.req_opcode);
-            payload[6] = reply.burst_complete ? 1 : 0;
-            put_le32_ptr(&payload[8], reply.offset);
-            memcpy(&payload[12], reply.data, sizeof(reply.data));
-            mavlink_msg_file_transfer_protocol_send(
-                reply.chan,
-                0, reply.sysid, reply.compid,
-                payload);
-            ftp.replies->pop();
-            ftp.last_send_ms = AP_HAL::millis();
-        } else {
-            return;
-        }
+    WITH_SEMAPHORE(comm_chan_lock(reply.chan));
```

Comment:
I might be missing something when I'm trying to audit the thread safety here. What stops this scenario from happening:

1. Take the lock here, and start sending an actual MAVLink message.
2. We get interrupted for the main flight code.
3. Main code does a notify on a parameter. I'm not seeing anything in this code path that results in us grabbing the lock back from the main channel. I fully believe I'm missing it, but can you point me at what stops this from happening?

Reply:
when the main thread does the param send, it calls mavlink_msg_param_value_send(). The generated mavlink headers then call MAVLINK_START_UART_SEND() which calls comm_send_lock(chan, size)
That causes the following sequence:
 - the main thread is suspended until it can get the semaphore
 - the FTP thread priority is raised to the priority of the main thread, which causes the FTP thread to be scheduled again, completing the send of the FILE_TRANSFER_PROTOCOL message and releasing the semaphore
 - the main thread then continues, allowing the param_value message send (if there is enough space in the buffer)

Reply:
note that we should consider adding a WITH_SEMAPHORE() in CHECK_PAYLOAD_SIZE(). It is not needed for this PR, but it would mean the payload space checks for other parts of the code would become more reliable. It will be quite a large patch as we'd need to add {} around all the switch statement branches in GCS_Common.cpp to limit the scope of the semaphore.